### PR TITLE
 Add restrictions and requirements of Proxy installation with Kuryr

### PR DESCRIPTION
**Pull request by MaysaMacedo**
Kuryr installation requires for a route entry to be added to the
Proxy host, and Kuryr defaults to always using http Proxy.

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1985486

Preview: https://deploy-preview-36621--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-restricted.html#installation-configure-proxy_installing-openstack-installer-restricted

### CODE REVIEW

File: modules/installation-configure-proxy.adoc
Comment (dulek) [inline modules/installation-configure-proxy.adoc:72]: I think in corporate environments the proxy will not run on bastion VM but will be some service managed by the corporate IT. So I'd make this more generic - simply the networking setup has to allow proxy to reply into the OpenShift cluster's router.

File: modules/installation-configure-proxy.adoc
Comment (dulek) [inline modules/installation-configure-proxy.adoc:74]: I'd capitalize HTTP.

**Review by dulek — CHANGES_REQUESTED** [review]
Maybe I'm misunderstanding something, I'm basing this on my general knowledge, not really OpenShift specifics.

File: modules/installation-configure-proxy.adoc
Comment (dulek) [inline modules/installation-configure-proxy.adoc:72]: I'm not sure if this is necessary now?

Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:72]: I left it there to be more explicit

Comment (dulek) [reply modules/installation-configure-proxy.adoc:72]: Fine with me if it's on purpose!

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [inline modules/installation-configure-proxy.adoc:47]: @MaysaMacedo Do you want this note to appear _only_ in [this document](https://deploy-preview-36621--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-custom?utm_source=github&utm_campaign=bot_dp), or should it appear in others, too?

Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:47]: @maxwelldb it should appear only in the openstack cluster-wide proxy related, so yeah I believe only that one

Comment (maxwelldb) [reply modules/installation-configure-proxy.adoc:47]: @MaysaMacedo Currently, this module is included in the doc that's linked in my previous comments, as well as this one: https://deploy-preview-36621--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-kuryr.html#installation-configure-proxy_installing-openstack-installer-kuryr

The `ifeval` configuration here means that your note _only_ appears in the linked document. 

Do you want it to appear:

- As it is right now
- In both docs
- In the latter doc (Kuryr) only

Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:47]: @maxwelldb it fit better being only in Kuryr

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [inline modules/installation-configure-proxy.adoc:75]: Should introduce the command explicitly, e.g. "To add a static route for the proxy configuration:"

**Review by maxwelldb — COMMENTED** [review]
Left a few more comments/questions.

File: modules/installation-configure-proxy.adoc
Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:77]: Yes, the PR that enforces this is already with lgtm and approve, just waiting on CI
https://github.com/openshift/cluster-network-operator/pull/1173/files#diff-6b64883c6488aec923e79dcf0febcaceda0c3d7f98a1a5832567e9455b3158eeR370-R373

File: modules/installation-configure-proxy.adoc
Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:75]: Good point, yes, better

**Comment by MaysaMacedo** [comment]
Dependent PR got merged https://github.com/openshift/cluster-network-operator/pull/1173

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [inline modules/installation-configure-proxy.adoc:71]:
```suggestion
For Kuryr installations on restricted networks that use the `Proxy` object, the proxy must be able to reply to the router of the {product-title} cluster. To add a static route for the proxy configuration:
```

Could "the proxy must be able to reply to the router of the {product-title} cluster" become "the proxy must be able to reply to the router that the cluster uses"? @MaysaMacedo

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [inline modules/installation-configure-proxy.adoc:75]:
```suggestion
Kuryr installations default to HTTP proxies.  
```

**Comment by maxwelldb** [comment]
@MaysaMacedo Which product versions does this apply to?

**Comment by MaysaMacedo** [comment]
> @MaysaMacedo Which product versions does this apply to?

It will probably go all the way to 4.6. It would make sense to backport when the dependent fix on cluster-network-operator is also backported.

File: modules/installation-configure-proxy.adoc
Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:71]: yes.

**Comment by MaysaMacedo** [comment]
@maxwelldb Thanks for the recommendations. I have squashed them into one commit.

**Comment by maxwelldb** [comment]
> It will probably go all the way to 4.6. It would make sense to backport when the dependent fix on cluster-network-operator is also backported.

@MaysaMacedo Sure. Could we get QE verification on that? Need to have commitments from QE specific to versions to go forward.

**Comment by MaysaMacedo** [comment]
> > It will probably go all the way to 4.6. It would make sense to backport when the dependent fix on cluster-network-operator is also backported.
> 
> @MaysaMacedo Sure. Could we get QE verification on that? Need to have commitments from QE specific to versions to go forward.
> 
> And to be clear, you mean:
> 
>     * 4.9: publish w/GA
> 
>     * 4.6-4.8: publish following the backport of the dependent CNO fix
> 
> 
> ?

yes. Regarding 4.9 I'm not sure if it would be published with GA since the code is in the process of getting backported.
I will contact QE.

**Comment by rlobillo** [comment]
During the verification work, we observed that there is an extra requirement on Kuryr: The default gateway of the restricted network in a proxied installation with Kuryr should be **defined and available**, as the router that Kuryr will create during its installation, will use it as one of its ports.

I believe it is worth to mention that here, just to avoid misunderstandings.

**Review by maxwelldb — COMMENTED** [review]
Added suggestions

**Review by maxwelldb — COMMENTED** [review]
@MaysaMacedo Does this seem copacetic to you?

**Comment by maxwelldb** [comment]
@rlobillo Can you give a last +1 to this? It's changed slightly since your last lgtm

**Comment by maxwelldb** [comment]
@MaysaMacedo Now that I look at this again, would it make sense as an item in the prerequisites list rather than as a note in the module description?

**Comment by MaysaMacedo** [comment]
> @MaysaMacedo Now that I look at this again, would it make sense as an item in the prerequisites list rather than as a note in the module description?

Yes, let me update that

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [inline modules/installation-configure-proxy.adoc:140]:
```suggestion
* For Kuryr installations on restricted networks that use the `Proxy` object, the proxy must be able to reply to the router that the cluster uses. To add a static route for the proxy configuration, from a command line, enter:
+
[source,terminal]
----
$ sudo ip route add <cluster_network_cidr> via <installer_subnet_gateway>
----

* The restricted subnet must have a gateway that is defined and available to be linked to the `Router` resource that Kuryr creates.
```

I believe this fixes a rendering issue in the prereq item. @MaysaMacedo

File: modules/installation-configure-proxy.adoc
Comment (michaelryanpeter) [inline modules/installation-configure-proxy.adoc:139]: Red Hat SSG recommends introducing a command "as the root user" over using `sudo`. So, the lead could be, "... from a command line, enter the following as a root user:"

Comment (michaelryanpeter) [reply modules/installation-configure-proxy.adoc:139]: > `$ sudo ip route add <cluster_network_cidr> via <installer_subnet_gateway>`
`# ip route add ... `

**Review by michaelryanpeter — COMMENTED** [review]
I left a small note regarding the SSG's guidelines on `sudo`. Otherwise, the PR looks good to me. Since I am a new peer review member, another member of the squad will follow up to check my review.

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [reply modules/installation-configure-proxy.adoc:139]: @MaysaMacedo Any objections to this change?

File: modules/installation-configure-proxy.adoc
Comment (maxwelldb) [inline modules/installation-configure-proxy.adoc:135]:
```suggestion
* For Kuryr installations on restricted networks that use the `Proxy` object, the proxy must be able to reply to the router that the cluster uses. To add a static route for the proxy configuration, from a command line as the root user, enter:
```

Comment (maxwelldb) [reply modules/installation-configure-proxy.adoc:135]: @MaysaMacedo

**Comment by maxwelldb** [comment]
@MaysaMacedo One more outstanding question for you ^^^

File: modules/installation-configure-proxy.adoc
Comment (MaysaMacedo) [reply modules/installation-configure-proxy.adoc:139]: I'm fine with it. Thanks for the suggestion

**Comment by maxwelldb** [comment]
@MaysaMacedo Thanks! I can merge this today. Still 4.6+?

**Comment by MaysaMacedo** [comment]
@maxwelldb Yes, but we only have merged 4.9 and 4.10 till now. The other backports are on the way.
Here is the open 4.8 one https://github.com/openshift/cluster-network-operator/pull/1225. Is there a way to include on the respective releases once the fix lands on those?

**Comment by maxwelldb** [comment]
So, because this is applicable to our `main` branch right now, I can merge it there. 

Because it's applicable to 4.9 and 4.9, I can cherry-pick to those release branches immediately after that. 

We can open an issue to cherry-pick to 4.6, 4.7, and 4.8 as appropriate--just need to reference this PR.

Does that sound good to you? @MaysaMacedo

**Comment by maxwelldb** [comment]
For reporting purposes, I'd just say that maybe the BZ should be split to reflect that the fixes are landing at different times, but from a docs perspective, this is just fixing a bug that impacts multiple released versions.

**Comment by MaysaMacedo** [comment]
@maxwelldb sounds good to me. We have a bugzilla for each version, but those point to cluster-network-operator. Should I open one for the docs as well?

**Comment by maxwelldb** [comment]
@MaysaMacedo Not sure it's necessary as long as those existing tickets link to the docs fix somehow. 

As I understand it, there's not a defect with the documentation--the docs changes are just triggered by code changes in this case. 

I just like traceability. :)

**Comment by MaysaMacedo** [comment]
@maxwelldb okay, I will make sure to either link the PR to the bz or add a comment to the BZs

**Comment by maxwelldb** [comment]
Note: PR CPs will be incomplete until additional changes are made to OCP 4.6-4.8. Made need to create new PR later depending on amount of change to docs set between now and future.

**Comment by maxwelldb** [comment]
@MaysaMacedo Thanks! That way, we can remember to cherrypick this change to 4.6-4.8 whenever that's good to go.

**Pull request merged**
